noir-projects/noir-contracts/contracts/protocol/tagging_contract/src/util.nr

```diff
@@ -0,0 +1,91 @@
+fn get_next_tag(
+    sender: AztecAddress,
+    recipient: AztecAddress,
+    hidden_sender: bool,
+    context: &mut PrivateContext,
+) -> Field {
+    if !handshake_exists(sender, recipient, hidden_sender) {
+        // The handshake doesn't exist, so we need to call the handshake function of the tagging contract.
+        context.call_private_function(
+            TAGGING_CONTRACT_ADDRESS,
+            FunctionSelector::from_signature("handshake((Field,bool))"),
+            [sender.to_field(), recipient.to_field(), hidden_sender as Field],
+        );
+    }
+
+    let master_tagging_public_key = get_master_tagging_public_key(sender, recipient, hidden_sender);
+
+    let sender_npk_m = get_public_keys(sender).npk_m;
+    let sender_nsk_app = context.request_nsk_app(sender_npk_m.hash());
+
+    // We call the is_first_time_tagging oracle function to determine if this is the first time tagging
+    // for [sender, recipient, hidden_sender] inputs. The function computes the tag nullifier with index
+    // 0 and checks if it exists in the nullifier tree.
+    let first_time_tagging: bool = is_first_time_tagging(sender, recipient, hidden_sender);
+
+    let (unsiloed_nullifier_preimage, app_to_silo_with, directional_app_tagging_secret, index) = if first_time_tagging {
+        // Sort addresses to reliably get bi-directional handshake commitment. We do that in an unconstrained function
+        // and then we constrain the result to optimize constraints.
+        let (address_0, address_1) = unsafe { sort_addresses(sender, recipient) };
+        assert(address_0.to_field() < address_1.to_field());
+        assert(address_0.eq(sender) | address_0.eq(recipient));
+        assert(address_1.eq(sender) | address_1.eq(recipient));
```

Comment:
As far as I remember Noir is not great at optimizing predicates so we could optimize this even more by moving this out of the if (then there would not be the predicate in the row of constraints or smt.).